### PR TITLE
Mesh coverage metric

### PR DESCRIPTION
**Pull request by mpiseno**
## Motivation and Context

In order to gauge how exhaustive a group of poses covers the entire mesh, this metric is needed. This metric uses the triangle sensor I made previously (https://github.com/facebookresearch/habitat-sim/pull/525) to count how many triangles of the entire mesh are covered by the group of poses determined from the image extractor. 

## How Has This Been Tested

I large part of this was figuring out how to get the total number of triangles in a semantic mesh, which I have verified is calculate correctly for mp3d scenes. I am working on more extensive testing and an opening this PR for feedback.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: habitat_sim/utils/data/data_extractor.py [hunk 1]
```diff
@@ -202,10 +205,37 @@ def get_semantic_class_names(self):
...
+            pos, rot, label, fp = pose
+            new_state = AgentState()
+            new_state.position = pos
+            new_state.rotation = rot
+            self.sim.agents[0].set_state(new_state)
+            obs = self.sim.get_sensor_observations()
+            triangle = obs["triangle_sensor"]
+            add_to_set(triangle)
+
+        total_num_triangles = self.sim._sim.get_num_triangles()
```
Comment (mathfac) [inline habitat_sim/utils/data/data_extractor.py]: I expect `get_num_triangles` should specify which mesh to use: RGB or Semantic.

File: habitat_sim/utils/data/data_extractor.py [hunk 2]
```diff
@@ -202,10 +205,37 @@ def get_semantic_class_names(self):
...
+        if len(self.scene_filepaths) > 1:
+            raise Exception(
+                "Percent mesh coverage metric is only \
+                valid on a single scene"
+            )
+
+        triangles_seen = set()
+        add_to_set = np.vectorize(lambda x: triangles_seen.add(x))
+        mymode = self.mode.lower()
+        poses = self.mode_to_data[mymode]
```
Comment (mathfac) [inline habitat_sim/utils/data/data_extractor.py]:
```suggestion
        poses = self.mode_to_data[self.mode.lower()]
```

File: src/esp/assets/ResourceManager.cpp [hunk 3]
```diff
@@ -1338,18 +1345,22 @@ void ResourceManager::loadMeshes(Importer& importer,
   int meshEnd = meshStart + importer.mesh3DCount() - 1;
   loadedAssetData.meshMetaData.setMeshIndices(meshStart, meshEnd);
 
+  // int totalNumTriangles = 0;
```
Comment (mathfac) [inline src/esp/assets/ResourceManager.cpp]: unintentional comments

**Comment by mathfac** [comment]
That's amazing seeing the mesh coverage works end-to-end. The unit test needed, also interested in values of the metric for current extraction algorithms.

**Comment by erikwijmans** [comment]
@mpiseno This should be computable for any given scene graph by just walking the drawable list (each drawable corresponds to a mesh and you can then easily get how many vertices are in the mesh: https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1Mesh.html#a578604bfa913ebe59b9b404b13d2a2af -- probably only makes sense to count the triangle meshes)

**Comment by eundersander** [comment]
Closing this stale PR.

**Pull request closed without merging**